crates/matrix-sdk-store-encryption/src/lib.rs

```diff
@@ -284,16 +284,83 @@ impl StoreCipher {
     /// # Result::<_, anyhow::Error>::Ok(()) };
     /// ```
     pub fn encrypt_value(&self, value: &impl Serialize) -> Result<Vec<u8>, Error> {
-        let mut data = serde_json::to_vec(value)?;
+        Ok(serde_json::to_vec(&self.encrypt_value_typed(value)?)?)
+    }
+
+    /// Encrypt a value before it is inserted into the key/value store.
+    ///
+    /// A value can be decrypted using the
+    /// [`StoreCipher::decrypt_value_typed()`] method. This is the lower
+    /// level function to `encrypt_value`, but returns the
+    /// full `EncryptdValue`-type
+    ///
+    /// # Arguments
+    ///
+    /// * `value` - A value that should be encrypted, any value that implements
+    /// `Serialize` can be given to this method. The value will be serialized as
+    /// json before it is encrypted.
+    ///
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # let example = || {
+    /// use matrix_sdk_store_encryption::StoreCipher;
+    /// use serde_json::{json, value::Value};
+    ///
+    /// let store_cipher = StoreCipher::new()?;
+    ///
+    /// let value = json!({
+    ///     "some": "data",
+    /// });
+    ///
+    /// let encrypted = store_cipher.encrypt_value_typed(&value)?;
+    /// let decrypted: Value = store_cipher.decrypt_value_typed(&encrypted)?;
+    ///
+    /// assert_eq!(value, decrypted);
+    /// # Result::<_, anyhow::Error>::Ok(()) };
+    /// ```
+    pub fn encrypt_value_typed(&self, value: &impl Serialize) -> Result<EncryptedValue, Error> {
```

Comment:
I'm don't really like the `_typed` name, but don't have any suggestions yet for how to change it.

Reply:
I agree, the naming isn't nice. But neither are the other options: either remove the suffix here and add a suffix to the base function (e.g. `_opaque`) but that imho gives the wrong one the higher visibility or give this a different suffix, but I don't really have any... Open for suggestions.